Volume/Supporting/AppDelegate.swift

```diff
@@ -6,21 +6,55 @@
 //  Copyright © 2021 Cornell AppDev. All rights reserved.
 //
 
+import AppDevAnalytics
 import UIKit
 import UserNotifications
 
-class AppDelegate: UIResponder, UIApplicationDelegate {
+class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
+
+    private let notificationIntervalKey = "notificationIntervalKey"
+
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
+        if let _ = launchOptions?[.remoteNotification] {
+            // app not running in background, user taps notification
+            AppDevAnalytics.log(VolumeEvent.clickNotification.toEvent(.notification, id: "", navigationSource: .pushNotification))
+            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: notificationIntervalKey)
+        }
         return true
     }
     
     func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
         let deviceTokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
         print("UIApplicationDelegate didRegisterForRemoteNotifications with deviceToken: \(deviceTokenString)")
+        print("this is the device token \(deviceToken.reduce("") { $0 + String(format: "%02.2hhx", $1) })")
         // send device token to backend
     }
     
     func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
         print("UIApplicationDelegate didRegisterForRemoteNotifications with error: \(error.localizedDescription)")
     }
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        // app running in background, user taps notification
+        AppDevAnalytics.log(VolumeEvent.clickNotification.toEvent(.notification, id: "", navigationSource: .pushNotification))
+        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: notificationIntervalKey)
+        completionHandler()
+    }
+
+    func applicationWillTerminate(_ application: UIApplication) {
+        let openDate = Date(timeIntervalSince1970: UserDefaults.standard.double(forKey: notificationIntervalKey))
```

Comment:
This will error if `UserDefaults` doesn't contain the initial timestamp if the user doesn't launch the app from a remote notification. You should probably handle the optional.